includes/utils/class-amp-image-dimension-extractor.php

```diff
@@ -47,73 +71,157 @@ public static function normalize_url( $url ) {
-			delete_transient( $transient_lock_name );
-			return false;
-		}
-
-		set_transient( $transient_name, $dimensions, $transient_expiry );
-		delete_transient( $transient_lock_name );
-		return $dimensions;
-	}
+    /**
+     * Determine which images to fetch by checking for dimensions in transient/cache.
+     * Creates a short lived transient that acts as a semaphore so that another visitor
+     * doesn't trigger a remote fetch for the same image at the same time.
+     *
+     * @param array $urls Urls of image src
+     * @param array $urls_to_fetch Urls of images to fetch because dimensions are not in transient/cache
+     * @param array $all_dimensions "Master" list of img url to dimension mappings - used here to track which dimensions couldn't be retrieved
+     */
+	private static function determine_which_images_to_fetch(&$urls, &$urls_to_fetch, &$all_dimensions) {
+        foreach ( $urls as $url ) {
+            $url_hash = md5( $url );
+            $transient_name = sprintf( 'amp_img_%s', $url_hash );
+            $dimensions = get_transient( $transient_name );
+
+            if ( is_array( $dimensions ) ) {
+                $all_dimensions[$url] = array(
+                    'width' => $dimensions[0],
+                    'height' => $dimensions[1],
+                );
+            } else {
+                if ( self::STATUS_FAILED_LAST_ATTEMPT === $dimensions ) {
+                    // If we couldn't get dimensions for this image the last time we tried, don't try again until the transient expires
+                    $all_dimensions[$url] = false;
+                    continue;
+                } else {
+                    // Ensure we're the only one trying to extract dimensions for this image at the moment so we don't duplicate requests
+                    $transient_lock_name = sprintf( 'amp_lock_%s', $url_hash );
+                    if ( false !== get_transient( $transient_lock_name ) ) {
+                        $all_dimensions[$url] = false;
+                        continue;
+                    } else {
+                        $urls_to_fetch[$url]['url'] = $url;
+                        $urls_to_fetch[$url]['transient_name'] = $transient_name;
+                        $urls_to_fetch[$url]['transient_lock_name'] = $transient_lock_name;
+                        set_transient($transient_lock_name, 1, MINUTE_IN_SECONDS);
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * Fetch dimensions of remote images
+     *
+     * @param array $urls_to_fetch Image src urls to fetch
+     * @param array $images Array to populate with results of image/dimension inspection
+     */
+    private static function fetch_images($urls_to_fetch, &$images, $mode) {
+        //User Faster Image when PHP version supports it (it contains a closure that could not be ported to 5.2)
+        if ( $mode === 'synchronous' || strnatcmp( phpversion(), '5.3.0' ) < 0 ) {
```

Comment:
I hadn't remembered doing anything specific to install multi curl after installing curl (I didn't need to install anything else). When Googling how to install multi curl didn't turn up any separate instructions from how to install curl, I performed a simple experiment to confirm that the availability of multi curl was tied to the availability of curl:

if ( function_exists('curl_exec') ) {
	echo "CURL EXISTS\n";
} else {
	echo "CURL DOES NOT EXIST\n";
}

if ( function_exists('curl_multi_exec') ) {
        echo "MULTI CURL EXISTS\n";
} else {
        echo "MULTI CURL DOES NOT EXIST\n";
}

Sure enough, when I disable the curl PHP extension multi curl becomes unavailable as well, and likewise when I re-enable it. Therefore, it doesn't seem like there are dependencies beyond curl.

That said, because curl itself is a dependency, and we don't want installations using php 5.3+ without curl to have errors, I've added a check for the existence of multi_curl_exec() and fall back to the original FastImage library if it isn't present.